### PR TITLE
End to end test!

### PR DESCRIPTION
**Pull request by paladine**
Also fixes memcpy and memcmp

### CODE REVIEW

File: MBBSEmu/DependencyInjection/ServiceResolver.cs [hunk 1]
```diff
@@ -30,6 +33,21 @@ public class ServiceResolver
         // Long term solution is probably to de-static this class :)
         private ServiceResolver() {}
 
+        public static void Create(IEnumerable<KeyValuePair<String,String>> data)
```
Comment (enusbaum) [inline MBBSEmu/DependencyInjection/ServiceResolver.cs:36]: `string` vs. the alias `String`

File: MBBSEmu/MBBSEmu.csproj [hunk 2]
```diff
@@ -11,6 +11,10 @@
     <None Remove="Assets\commandLineHelp.txt" />
     <None Remove="Assets\crashDetailsAPI.txt" />
     <None Remove="Assets\login.ans" />
+    <None Remove="Assets\MBBSEMU.DLL" />
```
Comment (enusbaum) [inline MBBSEmu/MBBSEmu.csproj:14]: We should move these to the Test project so they're not included in the embedded resources for the packaged executables

File: MBBSEmu/Session/TestSession.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            SessionType = EnumSessionType.Test;
+            SessionState = EnumSessionState.EnteringModule;
+        }
+
+        public override void Stop() {}
+
+        /// <summary>
+        ///     Reads data from the module until a new line is received, and returns the line with
+        ///     the line endings removed.
+        /// <param name="timeout">Maximum time to wait before throwing a TimeoutException</param>
```
Comment (enusbaum) [inline MBBSEmu/Session/TestSession.cs:30]: missing closing summary tag

File: MBBSEmu/Session/TestSession.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        public string GetLine(TimeSpan timeout)
+        {
+            return GetLine('\n', timeout).Trim('\r', '\n');
+        }
+
+        /// <summary>
+        ///     Reads data from the module until endingCharacter is received, and returns all data
+        ///     accumulated including endingCharacter
+        /// </summary>
+        /// <param name-"endingCharacter">Character which aborts reading</param>
```
Comment (enusbaum) [inline MBBSEmu/Session/TestSession.cs:40]: param name `=` vs `-`

File: MBBSEmu.Tests/Integration/Echo_Tests.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+using System.Text;
+using Xunit;
+
+namespace MBBSEmu.Tests.ExportedModules.Majorbbs
```
Comment (enusbaum) [inline MBBSEmu.Tests/Integration/Echo_Tests.cs:4]: move to `MBBSEmu.Tests.Integration` namespace and import `MBBSEmu.Tests.ExportedModules.Majorbbs`

File: MBBSEmu.Tests/Integration/MBBSEmuIntegrationTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+using MBBSEmu.HostProcess;
+using MBBSEmu.IO;
+using MBBSEmu.Module;
+using MBBSEmu.Resources;
+using MBBSEmu.Session;
+using System.Collections.Generic;
+using System.IO;
+using System;
+
+namespace MBBSEmu.Tests.ExportedModules.Majorbbs
```
Comment (enusbaum) [inline MBBSEmu.Tests/Integration/MBBSEmuIntegrationTestBase.cs:11]: move to MBBSEmu.Tests.Integration namespace and import MBBSEmu.Tests.ExportedModules.Majorbbs

File: MBBSEmu.Tests/Integration/MBBSEmuIntegrationTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                }
+            }
+        }
+
+        protected delegate void TestLogic(TestSession testSession);
+
+        protected void ExecuteTest(TestLogic testLogic)
+        {
+            var list = new List<KeyValuePair<string, string>>();
+            list.Add(new KeyValuePair<string, string>("BBS.Title", "Test"));
```
Comment (enusbaum) [inline MBBSEmu.Tests/Integration/MBBSEmuIntegrationTestBase.cs:61]: You can use an object initializer here, would remove the redundant `list.Add()`'s

File: MBBSEmu/DependencyInjection/ServiceResolver.cs (hunk 1 above)
Comment (paladine) [reply MBBSEmu/DependencyInjection/ServiceResolver.cs:36]: My Java is leaking again

File: MBBSEmu.Tests/Integration/Echo_Tests.cs (hunk 5 above)
Comment (paladine) [reply MBBSEmu.Tests/Integration/Echo_Tests.cs:4]: Copy/paste for the win!

File: MBBSEmu/Session/TestSession.cs [hunk 8]
```diff
@@ -79,7 +79,7 @@ public virtual void Send(byte[] dataToSend)
         /// <param name="dataToSend"></param>
         public void SendToModule(byte[] dataToSend)
         {
-            foreach(byte b in dataToSend)
+            foreach(var b in dataToSend)
```
Comment (enusbaum) [inline MBBSEmu/Session/TestSession.cs:83]: ❤️

**Review by enusbaum — APPROVED** [review]

**Pull request merged**